src/aioquic/tls.py

```diff
@@ -54,8 +54,17 @@
 
 T = TypeVar("T")
 
-# facilitate mocking for the test suite
-utcnow = datetime.datetime.utcnow
+# facilitate time mocking for the test suite
+#
+# Python 3.12 deprecates utcnow(), but we cannot do the recommended thing and replace
+# utcnow() with datetime.now(timezone.utc) as the datetime objects that the cryptography
+# x509 library puts in certificates do not have any timezone info.  If we try to
+# compare them with something that has timezone info we will get an exception:
+#
+#     TypeError: can't compare offset-naive and offset-aware datetimes
+#
+# We must continue to make an explicitly offset-naive timestamp.
+utcnow = partial(datetime.datetime.now, None)
```

Comment:
Hm, this doesn't look equivalent, I would have expected something like:

`datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None)`

Reply:
Looking into this a bit deeper it looks like we are going the wrong way here. We should be moving towards timezone-aware datetimes, as in:

- Making `utcnow = partial(datetime.datetime.now, datetime.timezone.utc)`
- Using `certificate.not_valid_before_utc` and `certificate.not_valid_after_utc`: https://cryptography.io/en/latest/x509/reference/#cryptography.x509.Certificate.not_valid_before_utc
- Requiring `cryptography >= 42.0.0`